File: service/runtimes/python311/bootstrap-websocket.py
#!/usr/bin/env python3

import json
import os
import sys
import time
import traceback
import asyncio
import websockets
import signal
from typing import Optional, Dict, Any

RUNTIME_API = os.environ.get('AWS_LAMBDA_RUNTIME_API', 'localhost:9001')
FUNCTION_NAME = os.environ.get('AWS_LAMBDA_FUNCTION_NAME')
FUNCTION_VERSION = os.environ.get('AWS_LAMBDA_FUNCTION_VERSION', '1')
MEMORY_SIZE = os.environ.get('AWS_LAMBDA_FUNCTION_MEMORY_SIZE')
LOG_GROUP_NAME = os.environ.get('AWS_LAMBDA_LOG_GROUP_NAME')
LOG_STREAM_NAME = os.environ.get('AWS_LAMBDA_LOG_STREAM_NAME')
INSTANCE_ID = os.environ.get('LAMBDAH_INSTANCE_ID')

# Load the user's handler
try:
    import lambda_function
    handler = lambda_function.handler
    if not callable(handler):
        raise Exception('Handler function not found in lambda_function.py')
except Exception as e:
    print(f'Failed to load handler: {e}', file=sys.stderr)
    sys.exit(1)

class WebSocketRuntime:
    def __init__(self):
        self.websocket: Optional[websockets.WebSocketServerProtocol] = None
        self.reconnect_attempts = 0
        self.max_reconnect_attempts = 10
        self.reconnect_delay = 1.0
        self.is_connected = False
        self.should_stop = False

    async def connect(self):
        """Connect to the WebSocket runtime API"""
        ws_url = f'ws://{RUNTIME_API}/2018-06-01/runtime/websocket?fn={FUNCTION_NAME}'
        print(f'Connecting to WebSocket: {ws_url}')

        try:
            self.websocket = await websockets.connect(ws_url)
            self.is_connected = True
            self.reconnect_attempts = 0
            print('WebSocket connected')
            await self.register()
            await self.message_loop()
        except Exception as e:
            print(f'WebSocket connection failed: {e}', file=sys.stderr)
            self.is_connected = False
            await self.handle_reconnect()

    async def register(self):
        """Register with the runtime API"""
        message = {
            'type': 'register',
            'function_name': FUNCTION_NAME,
            'runtime': 'python311',
            'version': FUNCTION_VERSION,
            'instance_id': INSTANCE_ID
        }
        await self.send(message)

    async def message_loop(self):
        """Main message processing loop"""
        try:
            async for message in self.websocket:
                try:
                    data = json.loads(message)
                    await self.handle_message(data)
                except json.JSONDecodeError as e:
                    print(f'Failed to parse WebSocket message: {e}', file=sys.stderr)
                except Exception as e:
                    print(f'Error handling message: {e}', file=sys.stderr)
        except websockets.exceptions.ConnectionClosed:
            print('WebSocket connection closed')
            self.is_connected = False
            await self.handle_reconnect()
        except Exception as e:
            print(f'WebSocket error: {e}', file=sys.stderr)
            self.is_connected = False
            await self.handle_reconnect()

    async def handle_message(self, message: Dict[str, Any]):
        """Handle incoming WebSocket messages"""
        message_type = message.get('type')
        
        if message_type == 'invocation':
            await self.handle_invocation(message)
        elif message_type == 'ping':
            await self.send({'type': 'pong'})
        elif message_type == 'error_response':
            print(f'Server error: {message.get("message", "Unknown error")}', file=sys.stderr)
        else:
            print(f'Unknown message type: {message_type}', file=sys.stderr)

    async def handle_invocation(self, message: Dict[str, Any]):
        """Handle function invocation"""
        request_id = message.get('request_id')
        payload = message.get('payload')
        deadline_ms = message.get('deadline_ms')
        invoked_function_arn = message.get('invoked_function_arn')
        trace_id = message.get('trace_id')
        
        print(f'Received invocation: {request_id}')

        try:
            # Create context object
            context = {
                'function_name': FUNCTION_NAME,
                'function_version': FUNCTION_VERSION,
                'memory_limit_in_mb': MEMORY_SIZE,
                'log_group_name': LOG_GROUP_NAME,
                'log_stream_name': LOG_STREAM_NAME,
                'aws_request_id': request_id,
                'invoked_function_arn': invoked_function_arn,
                'trace_id': trace_id
            }
            
            # Execute the user function
            result = handler(payload, context)
            
            # Send response
            await self.send({
                'type': 'response',
                'request_id': request_id,
                'payload': result,
                'headers': {
                    'X-Amz-Executed-Version': FUNCTION_VERSION
                }
            })
            
            print(f'Response posted for: {request_id}')
            
        except Exception as error:
            print(f'Handler error: {error}', file=sys.stderr)
            
            # Send error
            await self.send({
                'type': 'error',
                'request_id': request_id,
                'error_message': str(error),
                'error_type': type(error).__name__,
                'stack_trace': traceback.format_exc().split('\n'),
                'headers': {
                    'X-Amz-Function-Error': 'Unhandled'
                }
            })

    async def send(self, message: Dict[str, Any]):
        """Send a message via WebSocket"""
        if self.websocket and self.is_connected:
            try:
                await self.websocket.send(json.dumps(message))
            except Exception as e:
                print(f'Failed to send WebSocket message: {e}', file=sys.stderr)
        else:
            print('WebSocket not connected, cannot send message', file=sys.stderr)

    async def handle_reconnect(self):
        """Handle reconnection logic"""
        if self.reconnect_attempts < self.max_reconnect_attempts and not self.should_stop:
            self.reconnect_attempts += 1
            print(f'Attempting to reconnect ({self.reconnect_attempts}/{self.max_reconnect_attempts}) in {self.reconnect_delay}s')
            
            await asyncio.sleep(self.reconnect_delay)
            await self.connect()
            
            # Exponential backoff
            self.reconnect_delay = min(self.reconnect_delay * 2, 30.0)
        else:
            print('Max reconnection attempts reached, falling back to HTTP', file=sys.stderr)
            await self.fallback_to_http()

    async def fallback_to_http(self):
        """Fallback to HTTP runtime"""
        print('Falling back to HTTP runtime...')
        # Import and start the HTTP runtime
        import bootstrap
        # The HTTP runtime will start automatically when imported

    async def shutdown(self):
        """Graceful shutdown"""
        self.should_stop = True
        if self.websocket:
            await self.websocket.close()
        print('WebSocket runtime shutdown complete')

async def main():
    """Main entry point"""
    print('Lambda runtime started with WebSocket support')
    
    runtime = WebSocketRuntime()
    
    # Set up signal handlers for graceful shutdown
    def signal_handler(signum, frame):
        print(f'Received signal {signum}, shutting down...')
        asyncio.create_task(runtime.shutdown())
    
    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)
    
    try:
        await runtime.connect()
    except KeyboardInterrupt:
        print('Received keyboard interrupt, shutting down...')
        await runtime.shutdown()
    except Exception as e:
        print(f'Runtime error: {e}', file=sys.stderr)
        await runtime.shutdown()

if __name__ == '__main__':
    asyncio.run(main())

File: service/runtimes/nodejs18/bootstrap-websocket.js
#!/usr/bin/env node

const WebSocket = require('ws');
const fs = require('fs');
const path = require('path');

// Runtime API configuration (supports values with or without scheme)
const RAW_RUNTIME_API = process.env.AWS_LAMBDA_RUNTIME_API || 'host.docker.internal:9001';
function parseRuntimeApiHostPort(raw) {
  try {
    // Ensure URL has a scheme for URL parsing
    const url = raw.includes('://') ? new URL(raw) : new URL(`http://${raw}`);
    const hostname = url.hostname;
    const port = url.port ? parseInt(url.port, 10) : (url.protocol === 'https:' ? 443 : 80);
    return { hostname, port };
  } catch (e) {
    // Fallback to naïve split host:port
    const [host, p] = raw.split(':');
    return { hostname: host, port: p ? parseInt(p, 10) : 80 };
  }
}
const RUNTIME_API = parseRuntimeApiHostPort(RAW_RUNTIME_API);
const FUNCTION_NAME = process.env.AWS_LAMBDA_FUNCTION_NAME;
const FUNCTION_VERSION = process.env.AWS_LAMBDA_FUNCTION_VERSION || '1';
const HANDLER = process.env.AWS_LAMBDA_FUNCTION_HANDLER || 'index.handler';
const TASK_ROOT = process.env.LAMBDA_TASK_ROOT || '/var/task';
const INSTANCE_ID = process.env.LAMBDAH_INSTANCE_ID;

console.log('Lambda Runtime starting with WebSocket support...');
console.log('Function:', FUNCTION_NAME);
console.log('Handler:', HANDLER);
console.log('Task Root:', TASK_ROOT);

// Load the user function
let userHandler;
try {
    const handlerPath = path.join(TASK_ROOT, HANDLER.split('.')[0] + '.js');
    console.log('Loading handler from:', handlerPath);
    
    // Load the actual user function
    delete require.cache[require.resolve(handlerPath)];
    const userModule = require(handlerPath);
    const handlerName = HANDLER.split('.')[1] || 'handler';
    userHandler = userModule[handlerName];
    
    if (typeof userHandler !== 'function') {
        throw new Error(`Handler '${handlerName}' is not a function in ${handlerPath}`);
    }
    
    console.log('Successfully loaded handler:', handlerName);
} catch (error) {
    console.error('Failed to load handler:', error);
    process.exit(1);
}

// WebSocket message types
const MessageType = {
    REGISTER: 'register',
    INVOCATION: 'invocation',
    RESPONSE: 'response',
    ERROR: 'error',
    PING: 'ping',
    PONG: 'pong',
    ERROR_RESPONSE: 'error_response'
};

class WebSocketRuntime {
    constructor() {
        this.ws = null;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 10;
        this.reconnectDelay = 1000;
        this.isConnected = false;
    }

    connect() {
        const wsUrl = `ws://${RUNTIME_API.hostname}:${RUNTIME_API.port}/2018-06-01/runtime/websocket?fn=${encodeURIComponent(FUNCTION_NAME)}`;
        console.log('Connecting to WebSocket:', wsUrl);

        this.ws = new WebSocket(wsUrl);

        this.ws.on('open', () => {
            console.log('WebSocket connected');
            this.isConnected = true;
            this.reconnectAttempts = 0;
            this.register();
        });

        this.ws.on('message', (data) => {
            try {
                const message = JSON.parse(data.toString());
                this.handleMessage(message);
            } catch (error) {
                console.error('Failed to parse WebSocket message:', error);
            }
        });

        this.ws.on('close', (code, reason) => {
            console.log(`WebSocket closed: ${code} ${reason}`);
            this.isConnected = false;
            this.handleReconnect();
        });

        this.ws.on('error', (error) => {
            console.error('WebSocket error:', error);
            this.isConnected = false;
        });
    }

    register() {
        const message = {
            type: MessageType.REGISTER,
            function_name: FUNCTION_NAME,
            runtime: 'nodejs18.x',
            version: FUNCTION_VERSION,
            instance_id: INSTANCE_ID
        };
        this.send(message);
    }

    handleMessage(message) {
        switch (message.type) {
            case MessageType.INVOCATION:
                this.handleInvocation(message);
                break;
            case MessageType.PING:
                this.send({ type: MessageType.PONG });
                break;
            case MessageType.ERROR_RESPONSE:
                console.error('Server error:', message.message);
                break;
            default:
                console.warn('Unknown message type:', message.type);
        }
    }

    async handleInvocation(message) {
        const { request_id, payload, deadline_ms, invoked_function_arn, trace_id } = message;
        console.log('Got invocation:', request_id);

        try {
            // Execute the user function
            const result = await userHandler(payload);
            console.log('Function result:', result);

            // Send response
            this.send({
                type: MessageType.RESPONSE,
                request_id: request_id,
                payload: result,
                headers: {
                    'X-Amz-Executed-Version': FUNCTION_VERSION
                }
            });

            console.log('Posted response for:', request_id);
        } catch (error) {
            console.error('Function error:', error);
            
            // Send error
            this.send({
                type: MessageType.ERROR,
                request_id: request_id,
                error_message: error.message,
                error_type: 'Unhandled',
                stack_trace: error.stack ? error.stack.split('\n') : undefined,
                headers: {
                    'X-Amz-Function-Error': 'Unhandled'
                }
            });
        }
    }

    send(message) {
        if (this.ws && this.isConnected) {
            try {
                this.ws.send(JSON.stringify(message));
            } catch (error) {
                console.error('Failed to send WebSocket message:', error);
            }
        } else {
            console.warn('WebSocket not connected, cannot send message');
        }
    }

    handleReconnect() {
        if (this.reconnectAttempts < this.maxReconnectAttempts) {
            this.reconnectAttempts++;
            console.log(`Attempting to reconnect (${this.reconnectAttempts}/${this.maxReconnectAttempts}) in ${this.reconnectDelay}ms`);
            
            setTimeout(() => {
                this.connect();
            }, this.reconnectDelay);
            
            // Exponential backoff
            this.reconnectDelay = Math.min(this.reconnectDelay * 2, 30000);
        } else {
            console.error('Max reconnection attempts reached, falling back to HTTP');
            this.fallbackToHttp();
        }
    }

    fallbackToHttp() {
        console.log('Falling back to HTTP runtime...');
        // Import and start the HTTP runtime
        const httpRuntime = require('./bootstrap.js');
        // The HTTP runtime will start automatically when imported
    }
}

// Start WebSocket runtime
const runtime = new WebSocketRuntime();
runtime.connect();

// Handle graceful shutdown
process.on('SIGTERM', () => {
    console.log('Received SIGTERM, closing WebSocket connection');
    if (runtime.ws) {
        runtime.ws.close();
    }
    process.exit(0);
});

process.on('SIGINT', () => {
    console.log('Received SIGINT, closing WebSocket connection');
    if (runtime.ws) {
        runtime.ws.close();
    }
    process.exit(0);
});
